datacube/testutils/io.py

```diff
@@ -132,32 +137,168 @@ def native_geobox(ds: Dataset, measurements=None, basis: str | None = None):
+
+    :param ds: Sample dataset (only resolution and projection is used, not footprint)
+    :param band: Reference band to use
+    :param dst_geobox:
+                 (resolution of output GeoBox will match resolution of this band)
+    :param buffer: Buffer in units of CRS of ``ds`` (meters usually),
+                   default is 10 pixels worth
+    """
+    native: GeoBox = native_geobox(ds, basis=band)
+    if dst_geobox is None:
+        return native
+
+    if buffer is None:
+        buffer = 10 * max(map(abs, (native.resolution.y, native.resolution.x)))
+
+    assert native.crs is not None
+    return GeoBox.from_geopolygon(
+        dst_geobox.extent.to_crs(native.crs).buffer(buffer),
+        crs=native.crs,
+        resolution=native.resolution,
+        align=native.alignment,
+    )
 
-    :param ds: Dataset
-    :param measurements: List of band names to load
+
+def _split_by_grid(xx: xr.DataArray) -> list[xr.DataArray]:
+    """Split datasets by grid/crs"""
+
+    def extract(grid_id, ii):
+        yy = xx[ii]
+        crs = xx.grid2crs[grid_id]
+        yy.attrs.update(crs=crs)
+        yy.attrs.pop("grid2crs", None)
+        return yy
+
+    return [extract(grid_id, ii) for grid_id, ii in xx.groupby(xx.grid).groups.items()]
+
+
+def _native_load_1(
+    sources: xr.DataArray,
+    bands: tuple[str, ...],
+    *,
+    dst_geobox: GeoBox | None = None,
+    optional_bands: tuple[str, ...] | None = None,
+    basis: str | None = None,
+    pad: int | None = None,
+    **kw,
+) -> xr.Dataset:
+    """Load datasets with native crs
+    :param sources: grouped datasets
+    :param bands: List of band names to load
+    :param dst_geobox: Geobox of final output, if None then use geobox of input dataset
+    :param optional_bands: List of optional band names to load
     :param basis: Name of the band to use for computing reference frame, other
     bands might be reprojected if they use different pixel grid
+    :param pad: number of pixels to pad the geobox
 
-    :param kw: Any other parameter load_data accepts
+    :param kw: Any other parameter that load_data accepts
 
     :return: Xarray dataset
     """
-    from datacube import Datacube
-
-    geobox = native_geobox(
-        ds, measurements, basis
-    )  # early exit via exception if no compatible grid exists
-    mm = (
-        ds.product.measurements
-        if measurements is None
-        else ds.product.lookup_measurements(measurements)
-    )
+    if basis is None:
+        basis = bands[0]
+    (ds,) = sources.data[0]
+    load_geobox = compute_native_load_geobox(ds, basis, dst_geobox)
+    if pad is not None:
+        load_geobox = gbox_pad(load_geobox, pad)
+
+    mm = ds.product.lookup_measurements(bands)
+    if optional_bands is not None:
+        for ob in optional_bands:
+            try:
+                om = ds.product.lookup_measurements(ob)
+            except KeyError:
+                continue
+            else:
+                mm.update(om)
```

Comment:
Getting a KeyError exception should be preventable, would:
```python
measurements = ds.product.lookup_measurements(None)
for band in optional_bands:
    om = measurements.get(band)
    if om is not None:
        mm.update(om)
```
do the same thing?

Reply:
debugging/checking point for future change of mind/assumption

Reply:
I'm not a fan of leaving debugging/checkpoints in code long term, either.  Let's clean up.  Debugging/checkpoints can easily be reinstated temporarily when they are needed.

Reply:
All the debugging points are for revisiting after checking my current presumption on alchemist is valid, and fully compatible with stats afterwards.